lib/RestructureCFG/InlineDispatcherSwitch.cpp

```diff
@@ -737,13 +737,140 @@ inlineDispatcherSwitchImpl(ASTTree &AST,
   rc_return Node;
 }
 
-ASTNode *inlineDispatcherSwitch(ASTTree &AST, ASTNode *RootNode) {
+static RecursiveCoroutine<ASTNode *>
+simplifySuperfluousNonLocalCFImpl(ASTNode *Node,
+                                  const FallThroughScopeTypeMap
+                                    &FallThroughScopeMap) {
+  switch (Node->getKind()) {
+  case ASTNode::NK_List: {
+    SequenceNode *Seq = llvm::cast<SequenceNode>(Node);
+
+    // In place of a sequence node, we just need to inspect all the nodes in the
+    // sequence
+    bool ShouldSimplify = false;
+    for (ASTNode *&N : Seq->nodes()) {
+      if (not ShouldSimplify) {
+
+        // Compute the `FallThrough information after each `SequencenNode` is
+        // processed
+        N = rc_recur simplifySuperfluousNonLocalCFImpl(N, FallThroughScopeMap);
+        FallThroughScopeType ReturnType = FallThroughScopeType::Return;
+        ShouldSimplify = (FallThroughScopeMap.at(N) == ReturnType);
+      } else {
+        N = nullptr;
+      }
+    }
+
+    // In this beautify, it may be that a dispatcher switch is completely
+    // removed, therefore leaving a `nullptr` in a `SequenceNode`. We remove all
+    // these `nullptr` from the `SequenceNode` after the processing has taken
+    // place
+    Seq->removeNode(nullptr);
+
+    // This simplify should never completely empty a `SequenceNode`
+    revng_assert(not Seq->isEmpty());
```

Comment:
So, worst case, the sequence is left with a single node.

Don't we have an invariant that requires sequence composed of only 1 node to be unwrapped?

Reply:
You are right, but there are some caveats that we informally discussed in the past, briefly:
- In principle, we could run the three main steps to enforce all the invariants at each beautify step, by invoking the `normalize` method defined [here](https://github.com/revng/revng-c/blob/9624f2bbd14b6a3aead30f7c96fc6ed9a0b255ad/include/revng-c/RestructureCFG/GenerateAst.h#L1005), but this would mean traversing the AST multiple times.
- Therefore, we decided in the past to keep the passes that enforce a single invariant separately available.
-  And this is the case we are in. If you take a look a the beautify passes "pipeline" [here](https://github.com/revng/revng-c/blob/9624f2bbd14b6a3aead30f7c96fc6ed9a0b255ad/lib/RestructureCFG/BeautifyGHAST.cpp#L1025), you can see that the `simplifyAtomicSequence` pass (which is in charge to unwrap unitary sequences), is run after `inlineDispatcherSwitch`, therefore eventually enforcing the property you mentioned, with the assurance that with the current anatomy of the "pipeline" we do not have any other pass that requires such property enforced, between `inlineDispatcherSwitch` and `simplifyAtomicSequence` itself.

- We could, o.c., run it just after `inlineDispatcherSwitch` , and this would prevent possible problems with future reordering of the steps of the "pipeline", at the cost of visiting more times the AST. I think it could be worth discussing about this IRL if you want more details.

Reply:
Mmmh ok it makes sense like it is. Just note it in the comments.

Reply:
I added some comments commenting on the motivation for this assertion [here](https://github.com/revng/revng-c/blob/6af6c8ffa81d1d570010e6e64e260a78eedd4ea9/lib/RestructureCFG/RemoveDeadCode.cpp#L49)